### DISCUSSION TITLE
Device service throws "failed to find device <deviceName> in cache" ERROR

### DISCUSSION THREAD

**Discussion by sounak-1989** [edgexfoundry/community#19] [General]
Hi guys,

I am using Device Service SDK to create a new service for my SPI based devices (basically Input and output devices in SPI daisy chain). I am using the "Auto Discover" feature of the SDK.

After successful discovery my autoDiscover() returns the deviceModel structure which is then passed to the deviceCh channel which is of type []sdkModels.DiscoveredDevice.

I am following the Device LLRP as a reference for this design. But Whenever I run my service it shows that the devices are successfully added to metadata and I can query metadata to see the devices. But on the console logs it throws an error "failed to find device <deviceName> in cache". The source of this error is the manageddevice.go file.

Also when I am trying to send a command to my device I get the JSON response "device <deviceName> not found" although it shows the device when I query metadata. Not sure how this is happening. Since most of the stuff is happening inside the SDK I am not able to put breakpoints from within my device service project. 

Any help is much appreciated.

**Comment by lenny-goodell** [comment]
Check Core Metadata logs for error calling device service callback to Add Device.

This will occur if you are running in hybrid mode (device service from command line and the other EdgeX service in Docker) since your device service config has the Service Host set to `localhost`. Change the config to the IP of your system or simple stop and rerun the device service which will then find the device in Core Metadata at start-up.

**Reply by sounak-1989** [reply]
Thanks for the suggestion. I will try that out. Yes you are right. I am working in Hybrid mode. Do you think I need to register my device service with Consul? Thanks

**Reply by lenny-goodell** [reply]
Nope, it is just a hybrid dev thing you have to deal with since the device service is registering with Core Metadata with it host name set to `localhost` . In Docker `localhost` is the current container, not the host system, so Core Metadata can not get to your device service when it's host is set to `localhost`.
See here for more details: https://docs.edgexfoundry.org/2.2/getting-started/Ch-GettingStartedHybrid/

**Reply by sounak-1989** [reply]
Hi Lenny,
I tried what you suggested. From docker logs i see that the core-metadata is able to communicate with the device-service. The problem I am facing now is that whenever I try to send a GET or PUT command to any of the callbacks to read a resource or execute a command the HTTP server in the device-service goes into panic and spits the following messages.

2022/10/14 16:14:51 http: panic serving 10.30.4.26:60686: runtime error: invalid memory address or nil pointer dereference
goroutine 75 [running]:
net/http.(*conn).serve.func1()
	/usr/local/go/src/net/http/server.go:1850 +0x148
panic({0x110c200, 0x1766280})
	/usr/local/go/src/runtime/panic.go:890 +0x267
net/http.(*timeoutHandler).ServeHTTP(0xc00024e0c0, {0x12d8e18, 0xc00039a0e0}, 0xc000660200)
	/usr/local/go/src/net/http/server.go:3412 +0xa0e
github.com/gorilla/mux.(*Router).ServeHTTP(0xc00033fd40, {0x12d8e18, 0xc00039a0e0}, 0xc000660100)
	/home/sounak/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x25d
net/http.serverHandler.ServeHTTP({0xc000310000}, {0x12d8e18, 0xc00039a0e0}, 0xc0004c2000)
	/usr/local/go/src/net/http/server.go:2947 +0x474
net/http.(*conn).serve(0xc00061a000, {0x12d96d8, 0xc000372340})
	/usr/local/go/src/net/http/server.go:1991 +0x195c
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:3102 +0x9e5

Any idea what might be causing it? I can see that there is a nil pointer dereference but how do I debug it?

**Reply by cloudxxx8** [reply]
please attach the full log.

**Reply by sounak-1989** [reply]
Here's the full log from the device service:

level=INFO ts=2022-10-17T18:50:10.945379169Z app=device-io-hub source=config.go:391 msg="Loaded service configuration from ./res/configuration.toml"
level=INFO ts=2022-10-17T18:50:10.95292138Z app=device-io-hub source=config.go:551 msg="Using local configuration from file (0 envVars overrides applied)"
level=INFO ts=2022-10-17T18:50:10.953026353Z app=device-io-hub source=httpserver.go:131 msg="Web server starting (10.30.4.26:59999)"
level=INFO ts=2022-10-17T18:50:10.953067316Z app=device-io-hub source=messaging.go:69 msg="Setting options for secure MessageBus with AuthMode='usernamepassword' and SecretName='redisdb"
level=INFO ts=2022-10-17T18:50:10.95315414Z app=device-io-hub source=messaging.go:97 msg="Connected to redis Message Bus @ redis://localhost:6379 publishing on 'edgex/events/device' prefix topic with AuthMode='usernamepassword'"
level=INFO ts=2022-10-17T18:50:10.953206957Z app=device-io-hub source=init.go:134 msg="Check core-metadata service's status by ping..."
level=INFO ts=2022-10-17T18:50:10.957363625Z app=device-io-hub source=init.go:60 msg="Service clients initialize successful."
level=INFO ts=2022-10-17T18:50:10.957417613Z app=device-io-hub source=clients.go:124 msg="Using configuration for URL for 'core-metadata': http://localhost:59881"
level=INFO ts=2022-10-17T18:50:10.957448735Z app=device-io-hub source=restrouter.go:55 msg="Registering v2 routes..."
level=INFO ts=2022-10-17T18:50:10.961625875Z app=device-io-hub source=config.go:224 msg="Skipping use of Configuration Provider for custom configuration: Provider not available"
level=INFO ts=2022-10-17T18:50:10.961971134Z app=device-io-hub source=config.go:391 msg="Loaded custom configuration from ./res/configuration.toml"
level=INFO ts=2022-10-17T18:50:10.96205511Z app=device-io-hub source=config.go:283 msg="Loaded custom configuration from file (0 envVars overrides applied)"
level=INFO ts=2022-10-17T18:50:10.962079325Z app=device-io-hub source=simpledriver.go:142 msg="Custom config is: {5 1 res/provision_watchers}"
level=WARN ts=2022-10-17T18:50:10.962092987Z app=device-io-hub source=config.go:296 msg="unable to watch custom configuration for changes: Configuration Provider not enabled"
Error opening SPI port: spireg: no port found; did you forget to call Init()?
level=INFO ts=2022-10-17T18:50:10.963158212Z app=device-io-hub source=simpledriver.go:167 deviceName=IO-Hub-01 msg="Creating a new IO Card."
level=INFO ts=2022-10-17T18:50:10.963858905Z app=device-io-hub source=service.go:230 msg="device service device-io-hub exists, updating it"
level=INFO ts=2022-10-17T18:50:10.96502114Z app=device-io-hub source=profiles.go:54 msg="Loading pre-defined profiles from /home/sounak/work/fw-t3-device-services/cmd/device-io-hub/res/profiles"
level=INFO ts=2022-10-17T18:50:10.967129956Z app=device-io-hub source=profiles.go:88 msg="Profile digital-out exists, using the existing one"
level=INFO ts=2022-10-17T18:50:10.969639926Z app=device-io-hub source=profiles.go:88 msg="Profile input-card exists, using the existing one"
level=INFO ts=2022-10-17T18:50:10.9754283Z app=device-io-hub source=profiles.go:88 msg="Profile io-hub exists, using the existing one"
level=INFO ts=2022-10-17T18:50:10.975674786Z app=device-io-hub source=devices.go:49 msg="Loading pre-defined devices from /home/sounak/work/fw-t3-device-services/cmd/device-io-hub/res/devices"
level=INFO ts=2022-10-17T18:50:10.976027067Z app=device-io-hub source=devices.go:85 msg="Device IO-Hub-01 exists, using the existing one"
level=INFO ts=2022-10-17T18:50:10.97607362Z app=device-io-hub source=message.go:50 msg="Service dependencies resolved..."
level=INFO ts=2022-10-17T18:50:10.976090983Z app=device-io-hub source=message.go:51 msg="Starting device-io-hub 0.0.0 "
level=INFO ts=2022-10-17T18:50:10.97610087Z app=device-io-hub source=message.go:55 msg="IO Hub started"
level=INFO ts=2022-10-17T18:50:10.976110336Z app=device-io-hub source=message.go:58 msg="Service started in: 31.727141ms"
level=INFO ts=2022-10-17T18:50:10.976157702Z app=device-io-hub source=autodiscovery.go:51 msg="Starting auto-discovery with duration 1h0m0s"
level=INFO ts=2022-10-17T18:50:10.976176743Z app=device-io-hub source=simpledriver.go:754 msg="Discover was called."
level=INFO ts=2022-10-17T18:50:10.976188578Z app=device-io-hub source=simpledriver.go:643 msg="Adding provision watchers from res/provision_watchers"
level=INFO ts=2022-10-17T18:50:10.976382218Z app=device-io-hub source=discover.go:52 msg="Discovery completed. Device List size: 2"
level=INFO ts=2022-10-17T18:50:10.97640203Z app=device-io-hub source=discover.go:79 msg="New device: {IOCard6881339 digitalout 6881339 0 Digital Out Cards provide 16 single channel output or 8 differential digital outputs.}"
level=INFO ts=2022-10-17T18:50:10.976415523Z app=device-io-hub source=discover.go:79 msg="New device: {IOCard6160443 inputcard 6160443 1 Input Cards provide 8 single channel Voltage or 4 differential Voltage Inputs, 4 Current Inputs and 4 Digital Inputs.}"
level=INFO ts=2022-10-17T18:50:10.976428233Z app=device-io-hub source=discover.go:112 msg="Processing device: IOCard6881339."
level=ERROR ts=2022-10-17T18:50:10.976437673Z app=device-io-hub source=manageddevices.go:69 msg="failed to find Device IOCard6881339 in cache"
level=INFO ts=2022-10-17T18:50:10.976451395Z app=device-io-hub source=discover.go:146 msg="New device added: IOCard6881339"
level=INFO ts=2022-10-17T18:50:10.976458619Z app=device-io-hub source=discover.go:118 msg="New device created."
level=INFO ts=2022-10-17T18:50:10.976464488Z app=device-io-hub source=discover.go:112 msg="Processing device: IOCard6160443."
level=ERROR ts=2022-10-17T18:50:10.976470333Z app=device-io-hub source=manageddevices.go:69 msg="failed to find Device IOCard6160443 in cache"
level=INFO ts=2022-10-17T18:50:10.976479566Z app=device-io-hub source=discover.go:146 msg="New device added: IOCard6160443"
level=INFO ts=2022-10-17T18:50:10.976488294Z app=device-io-hub source=discover.go:118 msg="New device created."
level=INFO ts=2022-10-17T18:50:10.976497402Z app=device-io-hub source=simpledriver.go:807 msg="Discovered 2 new devices in 117.684µs."
level=INFO ts=2022-10-17T18:50:10.976550828Z app=device-io-hub source=async.go:104 msg="Adding discovered device IOCard6881339 to Metadata"
level=INFO ts=2022-10-17T18:50:10.978972863Z app=device-io-hub source=async.go:104 msg="Adding discovered device IOCard6160443 to Metadata"
2022/10/17 11:50:19 http: panic serving 10.30.4.26:36878: runtime error: invalid memory address or nil pointer dereference
goroutine 28 [running]:
net/http.(*conn).serve.func1()
        /usr/local/go/src/net/http/server.go:1850 +0xbf
panic({0xccaae0, 0x1436280})
        /usr/local/go/src/runtime/panic.go:890 +0x262
net/http.(*timeoutHandler).ServeHTTP(0xc00013a440, {0xf1b298, 0xc0003be380}, 0xc0001e4200)
        /usr/local/go/src/net/http/server.go:3412 +0x7ea
github.com/gorilla/mux.(*Router).ServeHTTP(0xc000347d40, {0xf1b298, 0xc0003be380}, 0xc0001e4000)
        /home/sounak/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x1cf
net/http.serverHandler.ServeHTTP({0xc000710090?}, {0xf1b298, 0xc0003be380}, 0xc0001e4000)
        /usr/local/go/src/net/http/server.go:2947 +0x30c
net/http.(*conn).serve(0xc0005e6000, {0xf1bbe0, 0xc000112c00})
        /usr/local/go/src/net/http/server.go:1991 +0x607
created by net/http.(*Server).Serve
        /usr/local/go/src/net/http/server.go:3102 +0x4db
^Clevel=INFO ts=2022-10-17T18:50:24.170132706Z app=device-io-hub source=httpserver.go:126 msg="Web server shutting down"
level=INFO ts=2022-10-17T18:50:24.170107572Z app=device-io-hub source=messaging.go:88 msg="Disconnected from MessageBus"
level=ERROR ts=2022-10-17T18:50:24.170232394Z app=device-io-hub source=httpserver.go:143 msg="Web server failed: http: Server closed"
level=INFO ts=2022-10-17T18:50:24.170423814Z app=device-io-hub source=httpserver.go:128 msg="Web server shut down"

**Reply by cloudxxx8** [reply]
Do you know where of the code opened a http server to listen to port 36878?
http: panic serving [10.30.4.26:36878](http://10.30.4.26:36878/)

According to the log, the default http server of the device service is listening to 59999

**Reply by sounak-1989** [reply]
Hi,
So I am sending the request from Postman to the device service. 59999 is the port where the device service is being hosted and [10.30.4.26:36878] is the ip and port from where postman is sending the request. Is it possible that the device service will crash if it receives request from any other service apart from the ones mentioned in the Client list?

I don't think that is the case since I am able to ping the device service from the same Postman request.

**Reply by sounak-1989** [reply]
Hi guys,

Figured out what the problem was. In my driver.go file I was trying to print a structure which is nil. Hence it throws the error and panics. This brings me to my initial problem. since none of the newly discovered devices are locally stored I am still getting the panic since my device service is not able to find the device locally. I will investigate further into it and keep you guys posted.

Thanks for all the help.

**Reply by cloudxxx8** [reply]
please check the log of core-metadata to confirm whether it can callback to the device service succesfully

**Reply by sounak-1989** [reply]
> please check the log of core-metadata to confirm whether it can callback to the device service successfully

Yes core-metadata can callback to the device. I can confirm this because when I try to delete the device through core metadata I can see the delete request coming to the device service and the device getting deleted. 
Although I have a question
- Is the device service supposed to create new local device files when auto discovery is turned on?

**Reply by cloudxxx8** [reply]
no, the container should be read-only, so there is no local files change.

in this case, why do you say the cache is not updated?  the device service cache are updated through the callback.  how did you confirm it?

**Reply by sounak-1989** [reply]
> no, the container should be read-only, so there is no local files change.
> 
> in this case, why do you say the cache is not updated? the device service cache are updated through the callback. how did you confirm it?

I am not running my device service from a container. I am working hybrid. So I was expecting the service to create new device files under cmd/res/devices/

I was initially asking about the cache because my device service was throwing an error which I mentioned on the topic. But now that issue is resolved once it properly started communicating with the other services.

Things are looking good now. The only problems are on my side of the driver code. I can see all the communications happening properly when I put breakpoints through the code.
